File: backend/app/schemas/gitops.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class GitOpsRepositoryCreate(BaseModel):
    """创建 GitOps 仓库配置的请求体。"""

    name: str = Field(min_length=1, max_length=100)
    repo_url: str = Field(min_length=1, max_length=500)
    branch: str = Field(default="main", min_length=1, max_length=100)
    local_path: str = Field(min_length=1, max_length=500)
    auth_type: str = Field(default="none", max_length=32)
    enabled: bool = True


class GitOpsRepositoryUpdate(GitOpsRepositoryCreate):
    """更新 GitOps 仓库配置的请求体。"""


class GitOpsRepositoryRead(BaseModel):
    """返回给前端展示的 GitOps 仓库配置。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    repo_url: str
    branch: str
    local_path: str
    auth_type: str
    enabled: bool
    last_sync_at: datetime | None = None
    last_commit_sha: str | None = None
    created_at: datetime
    updated_at: datetime


class GitOpsSyncRunRead(BaseModel):
    """返回给前端展示的 GitOps 同步记录。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    repository_id: int
    commit_sha: str | None = None
    status: str
    stdout: str | None = None
    stderr: str | None = None
    started_at: datetime
    finished_at: datetime | None = None


class DesiredResourceRead(BaseModel):
    """返回给前端展示的期望状态资源。"""

    id: int
    repository_id: int
    commit_sha: str
    resource_type: str
    resource_key: str
    file_path: str
    content: dict[str, Any]
    content_hash: str


class ActualResourceUpsert(BaseModel):
    """写入或更新 Actual Resource 的请求体。"""

    resource_type: str = Field(min_length=1, max_length=32)
    resource_key: str = Field(min_length=1, max_length=200)
    source: str = Field(default="manual", max_length=100)
    content: dict[str, Any]


class ActualResourceRead(BaseModel):
    """返回给前端展示的实际状态资源。"""

    id: int
    resource_type: str
    resource_key: str
    source: str
    content: dict[str, Any]
    content_hash: str
    scanned_at: datetime


class ResourceDiffRead(BaseModel):
    """返回给前端展示的资源差异。"""

    id: int
    repository_id: int
    resource_type: str
    resource_key: str
    diff_type: str
    before: dict[str, Any] | None = None
    after: dict[str, Any] | None = None
    risk_level: str
    created_at: datetime


class ApplyPlanRead(BaseModel):
    """返回给前端展示的执行计划。"""

    id: int
    repository_id: int
    diff_id: int
    plan: dict[str, Any]
    status: str
    policy_status: str
    ai_summary: str | None = None
    approved_by: int | None = None
    approved_at: datetime | None = None
    created_at: datetime


class PolicyResultRead(BaseModel):
    """返回给前端展示的策略校验结果。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    repository_id: int
    plan_id: int
    rule_key: str
    severity: str
    passed: bool
    message: str
    created_at: datetime

File: frontend/src/api/gitops.ts
import { getJson, postJson, putJson } from './client'

export interface GitOpsRepositoryPayload {
  name: string
  repo_url: string
  branch: string
  local_path: string
  auth_type: string
  enabled: boolean
}

export interface GitOpsRepository extends GitOpsRepositoryPayload {
  id: number
  last_sync_at?: string | null
  last_commit_sha?: string | null
  created_at: string
  updated_at: string
}

export interface GitOpsSyncRun {
  id: number
  repository_id: number
  commit_sha?: string | null
  status: string
  stdout?: string | null
  stderr?: string | null
  started_at: string
  finished_at?: string | null
}

export interface DesiredResource {
  id: number
  repository_id: number
  commit_sha: string
  resource_type: string
  resource_key: string
  file_path: string
  content: Record<string, unknown>
  content_hash: string
}

export interface ActualResourcePayload {
  resource_type: string
  resource_key: string
  source: string
  content: Record<string, unknown>
}

export interface ActualResource extends ActualResourcePayload {
  id: number
  content_hash: string
  scanned_at: string
}

export interface ResourceDiff {
  id: number
  repository_id: number
  resource_type: string
  resource_key: string
  diff_type: string
  before?: Record<string, unknown> | null
  after?: Record<string, unknown> | null
  risk_level: string
  created_at: string
}

export interface ApplyPlan {
  id: number
  repository_id: number
  diff_id: number
  plan: {
    resource_type: string
    resource_key: string
    diff_type: string
    steps: string[]
    rollback: string
  }
  status: string
  policy_status: string
  ai_summary?: string | null
  approved_by?: number | null
  approved_at?: string | null
  created_at: string
}

export interface PolicyResult {
  id: number
  repository_id: number
  plan_id: number
  rule_key: string
  severity: string
  passed: boolean
  message: string
  created_at: string
}

export interface GitOpsApplyRun {
  id: number
  repository_id: number
  plan_id: number
  stack_name: string
  target_path: string
  commit_sha?: string | null
  status: string
  stdout?: string | null
  stderr?: string | null
  raw_event_data?: string | null
  rollback_json: string
  started_at: string
  finished_at?: string | null
}
export function listGitOpsRepositories() {
  return getJson<GitOpsRepository[]>('/gitops-repositories')
}

export function createGitOpsRepository(payload: GitOpsRepositoryPayload) {
  return postJson<GitOpsRepository>('/gitops-repositories', payload)
}

export function updateGitOpsRepository(id: number, payload: GitOpsRepositoryPayload) {
  return putJson<GitOpsRepository>(`/gitops-repositories/${id}`, payload)
}

export function syncGitOpsRepository(id: number) {
  return postJson<GitOpsSyncRun>(`/gitops-repositories/${id}/sync`)
}

export function listGitOpsSyncRuns(id: number) {
  return getJson<GitOpsSyncRun[]>(`/gitops-repositories/${id}/sync-runs`)
}

export function listDesiredResources(id: number) {
  return getJson<DesiredResource[]>(`/gitops-repositories/${id}/desired-resources`)
}

export function upsertActualResource(payload: ActualResourcePayload) {
  return postJson<ActualResource>('/gitops-repositories/actual-resources', payload)
}

export function listActualResources() {
  return getJson<ActualResource[]>('/gitops-repositories/actual-resources')
}

export function generateDiffs(id: number) {
  return postJson<ResourceDiff[]>(`/gitops-repositories/${id}/diff`)
}

export function listDiffs(id: number) {
  return getJson<ResourceDiff[]>(`/gitops-repositories/${id}/diffs`)
}

export function listApplyPlans(id: number) {
  return getJson<ApplyPlan[]>(`/gitops-repositories/${id}/apply-plans`)
}

export function listPolicyResults(id: number) {
  return getJson<PolicyResult[]>(`/gitops-repositories/${id}/policy-results`)
}

export function approveApplyPlan(id: number) {
  return postJson<ApplyPlan>(`/gitops-repositories/apply-plans/${id}/approve`)
}

export function executeApplyPlan(id: number) {
  return postJson<GitOpsApplyRun>(`/gitops-repositories/apply-plans/${id}/execute`)
}

export function listApplyRuns(repositoryId: number) {
  return getJson<GitOpsApplyRun[]>(`/gitops-repositories/${repositoryId}/apply-runs`)
}
